### DISCUSSION TITLE
Am I using this correctly?

### DISCUSSION THREAD

**Discussion by mwasson74** [ardalis/GuardClauses#258] [Q&A] [answered]
Hello! Am I using this incorrectly or why does the compiler think that the return of `.Null` might be null? Is there a way to use it differently to let the compiler know that it won't be null when this returns? Is there a different extension I should be using? Thanks!!

```csharp
var sptf = Guard.Against.Null(sample.Sptfs.FirstOrDefault(x => x.Key == SptfKey), nameof(SptfKey), "I have no idea why this might be null");
var thisMightBeNull = sptf.Key;
```
![image](https://user-images.githubusercontent.com/12786751/219049566-28113000-62f7-442c-a9f8-180a5b618313.png)

**Comment by amal-stack** [comment, marked as answer]
@mwasson74 This is happening because the compiler does not have enough information to tell that `sptf.Key` will not be `null`  after calling `Guard.Against.Null()`. For now, you can safely suppress this warning using the null forgiving operator (`!`):
```c#
var thisMightBeNull = sptf!.Key
```

@ardalis This is possibly because `Guard.Against.Null` is defined as:
```c#
public static T Null<T>(this IGuardClause guardClause,
        [NotNull] [ValidatedNotNull] T input,
        [CallerArgumentExpression("input")] string? parameterName = null,
        string? message = null)
```
If the `input` parameter definition is changed from:
```c#
T input
```
to
```c#
T? input
```
the warning mentioned in the question goes away. 
Since the type `T` does not have a constraint like `struct` or `class`, the `T?` syntax will only be supported from C# 9. To use the `T?` syntax prior to C# 9, there must exist a `class` or `struct` constraint for the type `T`. The pre-C# 9 equivalent to `T?` seems to be the `AllowNullAttribute`.  

**Edit**: PR #261 which fixes this issue has been merged. The fix is available since version *4.1.0*.

**Reply by mwasson74** [reply]
Thanks, @amal-stack, we ended up using the null-forgiving operator here. I knew I could safely use it but assumed I just wasn't using .Null correctly. Hopefully a small update/PR will allow this to work as I assumed it did.
